engines/director/lingo/xlibs/spacemgr.cpp

```diff
@@ -0,0 +1,670 @@
+	if (type == kXObj) {
+		SpaceMgrXObject::initMethods(xlibMethods);
+		SpaceMgrXObject *xobj = new SpaceMgrXObject(kXObj);
+		g_lingo->exposeXObject(xlibName, xobj);
+	} else if (type == kXtraObj) {
+		// TODO - Implement Xtra
+	}
+}
+
+void SpaceMgr::close(int type) {
+	if (type == kXObj) {
+		SpaceMgrXObject::cleanupMethods();
+		g_lingo->_globalvars[xlibName] = Datum();
+	} else if (type == kXtraObj) {
+		// TODO - Implement Xtra
+	}
+}
+
+void SpaceMgr::m_new(int nargs) {
+	if (nargs != 0) {
+		warning("SpaceMgr::m_new: expected 0 arguments");
+		g_lingo->dropStack(nargs);
+	}
+	g_lingo->push(g_lingo->_currentMe);
+}
+
+void SpaceMgr::m_dispose(int nargs) {
+	g_lingo->printSTUBWithArglist("SpaceMgr::m_dispose", nargs);
+	g_lingo->dropStack(nargs);
+}
+
+void SpaceMgr::m_lastError(int nargs) {
+	g_lingo->printSTUBWithArglist("SpaceMgr::m_lastError", nargs);
+	g_lingo->dropStack(nargs);
+	g_lingo->push(Datum(0));
+}
+
+void SpaceMgr::m_memUsed(int nargs) {
+	g_lingo->printSTUBWithArglist("SpaceMgr::m_memUsed", nargs);
+	g_lingo->dropStack(nargs);
+	g_lingo->push(Datum(0));
+}
+
+void SpaceMgr::m_listData(int nargs) {
+	g_lingo->printSTUBWithArglist("SpaceMgr::m_listData", nargs);
+	g_lingo->dropStack(nargs);
+	g_lingo->push(Datum(""));
+}
+
+void SpaceMgr::m_sortAll(int nargs) {
+	g_lingo->printSTUBWithArglist("SpaceMgr::m_sortAll", nargs);
+	g_lingo->dropStack(nargs);
+	g_lingo->push(Datum(0));
+}
+
+void SpaceMgr::m_checkForDups(int nargs) {
+	if (nargs != 1) {
+		warning("SpaceMgr::m_checkForDups: expected 1 argument");
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Datum(0));
+		return;
+	}
+	Datum arg = g_lingo->pop();
+	if (arg.type != STRING) {
+		warning("SpaceMgr::m_checkForDups: expected arg to be a string, not %s", arg.type2str());
+		g_lingo->push(Datum(0));
+		return;
+	}
+
+	SpaceMgrXObject *me = static_cast<SpaceMgrXObject *>(g_lingo->_currentMe.u.obj);
+
+	me->_checkForDups = (arg.u.s->c_str()[0] == 't') || (arg.u.s->c_str()[0] == 'T');
+	g_lingo->push(Datum(0));
+}
+
+void SpaceMgr::m_parseText(int nargs) {
+	if (nargs != 1) {
+		warning("SpaceMgr::m_parseText: expected 1 argument");
+		g_lingo->dropStack(nargs);
+		g_lingo->push(Datum(0));
+		return;
+	}
+	Datum text = g_lingo->pop();
+	if (text.type != STRING) {
+		warning("SpaceMgr::m_parseText: expected text to be a string, not %s", text.type2str());
+		g_lingo->push(Datum(0));
+		return;
+	}
+	SpaceMgrXObject *me = static_cast<SpaceMgrXObject *>(g_lingo->_currentMe.u.obj);
+
+	Common::String result = *text.u.s;
+	Common::String curSpaceCollection;
+	Common::String curSpace;
+	Common::String curNode;
+	Common::String curView;
+	int lastIndex = 0;
+	while (lastIndex < (int)result.size()) {
+		size_t index = result.findFirstOf('\r', lastIndex);
+		if (index == Common::String::npos)
+			index = result.size();
+		int pointer = lastIndex;
+
+		while (pointer < (int)index && result[pointer] == ' ')
+			pointer++;
+		if ((int)result.find("SPACECOLLECTION ", pointer) == pointer) {
```

Comment:
That is a bit of an interesting approach here. I would personally use `Common::StringTokenizer`, but of course, it is up to the individual taste.

Reply:
I'm open for refactoring once the prototyping is done